service/topology_coordinator.cc

```diff
@@ -850,6 +853,90 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                const table_id& table_id = topo.truncate_table_id.value();
+                const std::chrono::milliseconds timeout = topo.truncate_table_timeout_ms.value();
+                lw_shared_ptr<replica::table> table = _db.get_tables_metadata().get_table_if_exists(table_id);
+
+                std::optional<sstring> error;
+                if (table) {
+                    const sstring& ks_name = table->schema()->ks_name();
+                    const sstring& cf_name = table->schema()->cf_name();
+
+                    rtlogger.info("Performing TRUNCATE TABLE global topology request for {}.{} with timeout {}", ks_name, cf_name, timeout);
+
+                    const locator::tablet_map& tmap = get_token_metadata_ptr()->tablets().get_tablet_map(table_id);
+
+                    // Collect the IPs of the hosts with replicas
+                    std::unordered_set<gms::inet_address> replica_endpoints;
+                    co_await tmap.for_each_tablet([&] (locator::tablet_id tid, const locator::tablet_info& tinfo) {
+                        for (const locator::tablet_replica& replica: tinfo.replicas) {
+                            const std::set<gms::inet_address> addrs = _gossiper.get_nodes_with_host_id(replica.host);
+                            replica_endpoints.insert(addrs.begin(), addrs.end());
+                        }
+                        return make_ready_future<>();
+                    });
+
+                    // Check if all the nodes with replicas are alive
+                    const std::set<gms::inet_address> live_nodes = _gossiper.get_live_members();
+                    const size_t live_replicas = std::count_if(replica_endpoints.begin(), replica_endpoints.end(),
+                                                                [&live_nodes] (const gms::inet_address& addr) {
+                                                                    return live_nodes.contains(addr);
+                                                                });
+                    if (live_replicas == replica_endpoints.size()) {
+                        try {
+                            // Send the RPC to all replicas
+                            const netw::messaging_service::clock_type::time_point timeout_deadline = lowres_clock::now() + timeout;
+                            const service::frozen_topology_guard frozen_guard { _topo_sm._topology.session };
+                            co_await coroutine::parallel_for_each(replica_endpoints, [&] (const gms::inet_address& ep) -> future<> {
+                                netw::messaging_service::msg_addr addr(ep);
+                                co_await ser::storage_proxy_rpc_verbs::send_truncate(&_messaging, addr, timeout_deadline, ks_name, cf_name, frozen_guard);
+                            });
+                        } catch (std::exception& e) {
+                            error = e.what();
```

Comment:
So basically failure may mean that data was partially deleted, right?

Reply:
should we propagate a timeout error back to the coordinator of the truncate cmd? Also, I see the attempt to truncate all happens in a while(true) loop, but I have a feeling we should respect the deadline established by topo.truncate_table_timeout_ms. there's no need to retry if we're past the deadline since the beginning of the request, no?

Reply:
> So basically failure may mean that data was partially deleted, right?

Yes, that is right. AFAIK that is how the current truncate implementation works, too.

Reply:
> should we propagate a timeout error back to the coordinator of the truncate cmd?

I think we should, and that is what we do by serializing the error string into `system.topology_requests.error`. It is then read in `storage_service::truncate_table_with_tablets()` and thrown from there. The problem is that the exception type thrown back to the cmd will always be `std::runtime_error` instead what was actually caught in the topology coordinator.

> Also, I see the attempt to truncate all happens in a while(true) loop, but I have a feeling we should respect the deadline established by topo.truncate_table_timeout_ms. there's no need to retry if we're past the deadline since the beginning of the request, no?

You are right about respecting the timeout. Actually, I think we should create a deadline time point the moment we receive the TRUNCATE command, and serialize that as a parameter in `system.topology` instead of `topo.truncate_table_timeout_ms`, and then use that deadline in truncate RPC calls. That way, the timeout will also respect the time it took to queue the topology request. What do you think?

Reply:
Typically creating a deadline is the right thing, but with truncate, there's not much a user can do with a partially complete truncate, or with a truncate that didn't execute because it missed the deadline. Both are an indication of a broken system. 

Reasons for truncate timing out:
 - it's naturally slow, the driver isn't aware that it's a truncate operation, and there's no way to send keep-alives; this will take much more effort to solve
 - the CQL coordinator died (but here it's sort-of-reasonable)
 - one of the nodes is down or partitioned

I suggest to avoid the timeout/deadline beyond the CQL coordinator level. After all we don't have a deadline for schema changes.

Reply:
> I suggest to avoid the timeout/deadline beyond the CQL coordinator level. After all we don't have a deadline for schema changes.

If I understand correctly, that would mean that the truncate RPC needs to be changed so that it does not have a timeout:
https://github.com/scylladb/scylladb/blob/8fb6d98ba3e6b8ff04c06c462325a126bdcec4f8/idl/storage_proxy.idl.hh#L34

How does removing that attribute affect older nodes in a cluster (ones that still have this IDL attribute)? Maybe we could introduce a new RPC and leave the old one for the current truncate implementation? 

Or, we could call `ser::storage_proxy_rpc_verbs::send_truncate()` with a deadline very far into the future, but that does not seem like a very nice solution.

Reply:
I don't think with_timeout affects the wire format. In any case you could keep it, and send it with a large timeout.I don't know what's reasonable, could put 15 minutes there. I don't think we really need a timeout - the failure detector, and tcp keepalive, will tear down connections to non-responsive nodes.

Reply:
btw you could add a separate truncate handler.

Reply:
The coordinator should not satisfy the request / update the topology state machine on failure.  If the client's side of truncate times out, this will leave the truncate status uncertain to the client, and to an apparent data resurrection. 
I think retrying (unless the error is irrecoverable) is a safer strategy.

Reply:
Retrying on recoverable errors has been implemented in v8.